proposal.md

```diff
@@ -0,0 +1,1451 @@
++ func AppendUnquote[Bytes ~[]byte | ~string](dst []byte, src Bytes) ([]byte, error)
+```
+
+The length returned by `StackIndex` is now a `int64` instead of `int`
+since the length of a JSON array or object could theoretically exceed `int`
+when handling JSON in a purely streaming manner.
+The stack depth, however, remains fundamentally limited by
+the amount of system memory, so an `int` is still an appropriate type.
+```diff
+-func (e *Encoder) StackIndex(i int) (Kind, int)
++func (e *Encoder) StackIndex(i int) (Kind, int64)
+-func (d *Decoder) StackIndex(i int) (Kind, int)
++func (d *Decoder) StackIndex(i int) (Kind, int64)
+```
+
+The pointer returned by `StackPointer` is now the named `Pointer` type.
+```diff
+- func (e *Encoder) StackPointer() string
++ func (e *Encoder) StackPointer() Pointer
+- func (d *Decoder) StackPointer() string
++ func (d *Decoder) StackPointer() Pointer
+```
+
+An explicit `Pointer` type was added to represent a JSON Pointer
+([RFC 6901](https://www.rfc-editor.org/rfc/rfc6901))
+as a means to identify exactly where an error occurred.
+Convenience methods are defined for interacting with a pointer.
+```diff
++ type Pointer string
++ func (p Pointer) IsValid() bool
++ func (p Pointer) AppendToken(tok string) Pointer
++ func (p Pointer) Parent() Pointer
++ func (p1 Pointer) Contains(p2 Pointer) bool
++ func (p Pointer) LastToken() string
++ func (p Pointer) Tokens() iter.Seq[string]
+```
+
+Handling of errors was improved:
+```diff
+  type SyntacticError struct {
+  	...
+- 	JSONPointer string
++ 	JSONPointer Pointer
+  }
+
++ var ErrDuplicateName = errors.New("duplicate object member name")
++ var ErrNonStringName = errors.New("object member name must be a string")
+```
+
+The `ErrDuplicateName` and `ErrNonStringName` errors were added to support
+common error conditions users may want to distinguish upon
+through the use of `errors.Is`.
+
+### Package "encoding/json/v2"
+
+Interface types and methods were renamed to avoid the `V1` and `V2` suffixes,
+which were aesthetically unpleasant. Instead, the `V2` declarations now
+generally use the `To` and `From` suffixes to indicate that they support streaming.
+This follows after the convention established by `io.WriterTo` and `io.ReaderFrom`.
+```diff
+- type MarshalerV1 interface { MarshalJSON() ([]byte, error) }
++ type Marshaler interface { MarshalJSON() ([]byte, error)}
+
+- type MarshalerV2 interface { MarshalJSONV2(*jsontext.Encoder, Options) error }
++ type MarshalerTo interface { MarshalJSONTo(*jsontext.Encoder, Options) error}
+
+- type UnmarshalerV1 interface { UnmarshalJSON([]byte) error }
++ type Unmarshaler interface { UnmarshalJSON([]byte) error}
+
+- type UnmarshalerV2 interface { UnmarshalJSONV2(*jsontext.Decoder, Options) error }
++ type UnmarshalerFrom interface { UnmarshalJSONFrom(*jsontext.Decoder, Options) error}
+
+- func MarshalFuncV1[T any](fn func(T) ([]byte, error)) *Marshalers
++ func MarshalFunc[T any](fn func(T) ([]byte, error)) *Marshalers
```

Comment:
thinking out of the box for a second here - Marshalers is a new feature, and we want to encourage users to use the new marshaler interface which is more efficient - should we be exposing a Marshalers API for the old interface signature?

I guess it can be useful for a transition period where many types expose the old methods but not the new ones yet, but after a couple of years, I imagine this will be rather pointless.

Reply:
I vaguely recall us having this discussion and transition is part of the reason. We'll want people to be able to combine `json.MarshalFunc` with `protojson.Marshal`, that is until protobuf implements `protojson.MarshalEncode`.